templates/tfengine/components/devops/main.tf

```diff
@@ -89,17 +83,21 @@ module "owners_group" {
   source  = "terraform-google-modules/group/google"
   version = "~> 0.1"
 
-  id = "{{.project.owners_group.id}}"
-  customer_id = "{{.project.owners_group.customer_id}}"
-  {{- if has .project.owners_group "display_name"}}
-  display_name = "{{.project.owners_group.display_name}}"
-  {{- else}}
-  display_name = "{{regexReplaceAll "@.*" .project.owners_group.id ""}}"
+  id = var.project.owners_group.id
+  customer_id = var.project.owners_group.customer_id
+  display_name = var.project.owners_group.display_name
+  {{- if has .project.owners_group "description"}}
+  description = var.project.owners_group.description
+  {{- end}}
+  {{- if has .project.owners_group "owners"}}
+  owners = var.project.owners_group.owners
+  {{- end}}
+  {{- if has .project.owners_group "managers"}}
+  managers = var.project.owners_group.managers
+  {{- end}}
+  {{- if has .project.owners_group "members"}}
+  members = var.project.owners_group.members
```

Comment:
Unfortunately, `hclFIeld` doesn't work here. We're trying to insert something like `var.admin_group.description`, but `hclField` inserts the actual value.

It thought to create an `hclFIeldToVar(m map[string]interface{}, key string)` but it has to assume that:

1. The string variable that receives is in a nested object at the `variables.tf`, and not necessarily. How does it decides if the variable should be written as `var.admin_group.description` or `var.admin_group_description`? The resolution for this will be completely dependant on how private catalog handles nested objects, and also, it conflicts with the following [suggestion](https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/pull/908#discussion_r657367620):

>2) If not too difficult, add them to the main variable's description, on a separate line, since this way they show up when terraform prints information about the variable (comments are ignored so a person would have to open the variables.tf file)

If we decide to take the quoted approach, we should define a standard way to convert nested objects to main variables. And that's too difficult to keep track of IMO

2. There should be a way to get the name of the map. How does the function know that the maps is actually called `admins_group` to append it in `var.admin_group.description`?. The only way I see around it is to specify it as a parameter. But again, it's too hard to keep track of.

In conclusion, I don't see a better way to render variable equivalents if the field is present. Wyt?

Reply:
I see. I just discovered this, why not use the [try function](https://www.terraform.io/docs/language/functions/try.html)? It would look like this:

```
description = try(var.project.owners_group.description, null)
owners = try(var.project.owners_group.owners, null)
managers = try(var.project.owners_group.managers, null)
members = try(var.project.owners_group.members, null)
```

Reply:
Love it, but now testing it, it makes the variables required at the `variables.tf`. I can replace the `hclField`s there, but that would require me to add empty arrays.

Insted of:
`{{hclField .admins_group "owners" -}}`
doing:
`owners = {{get .admins_group "description" "[]"}}`

What do you think is better?

First option we're not able to use `try` and should keep the `ifs`
Second option we'll be rendering empty arrays, which I don't know if they have the same effect on terraform and also exposes fields that are not even supported at the moment in the schema docs (managers and members)

Reply:
Continuation for https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/pull/908#issuecomment-870567206

>I can't find this comment in the UI anymore so I'll respond via email.
>
>To my understanding, only the "project" variable is now required, all its
>sub-values may or may not exist. Then, the variable can be defined in
>variables.tf but given a default value of "null" (which works for all
>variable types). Or are you talking about how this would affect other
>places where it's used? If so, I also think it's okay to have the variable
>defined there and given a null default value if necessary.

I had several problems using the `try` approach. It's not that the variable might be present or not, it's that it's declared and just because it's used inside the `try`, it should be added to `terraform.tfvars`

I talked with Xin during on our 1-1, and suggested to add default values. But again, many problems.

Now the problem is that defining defaults for objects assume that is the default for the whole object. For example:

```
// variables.tf
...
var = {
  a = string
  b = list(string)
}
default = {
  b = []
}
```

Would throw error because it needs to declare an `a` default too.
Also, with:

```
// variables.tf
...
var = {
  a = string
  b = list(string)
}
default = {
  a = ""
  b = []
}
```

Doesn't work because if you only specify `a`, it the specified variable is not merged with the default. I mean, it doesn't put only the `b` default. If the variable is provided but no `b`, the default has no effect.

I'm working on a doc with my notes regarding known problems that I've found, I just have to put it in order because there are too many cases, so I can share it with you because I'm having troubles explaining myself in general. 

So far, the best approach that I found is to use `if` statements to conditionally render the variables in `variables.tf`, `terrfaform.tfvars` and `main.tf`, which is difficult to track.

Reply:
> Doesn't work because if you only specify a, it the specified variable is not merged with the default. I mean, it doesn't put only the b default. If the variable is provided but no b, the default has no effect.

Didn't fully get this part, but I'll wait for your doc.

Reply:
Pretty simple, if you have

```
variable "foo" {
  type = object({
    a = string
    b = list(string)
  })
  default = {
    a = ""
    b = []
  }
}
```

And you specify this at `terraform.tfvars`

```
foo = {
  a = "bar"
}
```

The final result will be

```
foo = {
  a = "bar"
}
```

I mean that foo doesn't appear because the default apply as a whole, and not merge defaults with specified variables.

Reply:
Got it. Does https://www.terraform.io/docs/language/functions/defaults.html help at all?

Reply:
Another way is define the type to be `any`, and do `lookup`s in the `main.tf`.

https://github.com/terraform-google-modules/terraform-google-healthcare/blob/master/main.tf#L34
https://github.com/terraform-google-modules/terraform-google-healthcare/blob/master/variables.tf#L62

The problem of this again is I'm not sure how these will surface in Private Catalog's UI.

Reply:
> Got it. Does https://www.terraform.io/docs/language/functions/defaults.html help at all?

Quickly tested and works, but now I have to define values as `optional()` or not, which is almost the same logic, and it doesn't fully solve the same problem at `terraform.tfvars`. I have to conditionally render there, and since the conditions are pretty much the same to conditionally render at `variables.tf`, I see more suitable the solution that I already got (Note unit tests are passing).

> Another way is define the type to be `any`, and do `lookup`s in the `main.tf`.
> 
> https://github.com/terraform-google-modules/terraform-google-healthcare/blob/master/main.tf#L34
> https://github.com/terraform-google-modules/terraform-google-healthcare/blob/master/variables.tf#L62
> 
> The problem of this again is I'm not sure how these will surface in Private Catalog's UI.

This one seems interesting. Let me try it. I'll just open another branch against this one to test because I don't wanna mix what I already have

Reply:
I'm ok with go with the current approach for now as I don't want us to spend too much time on this without understanding more about Private Catalog and whether or not this approach will actually work with other recipes. Things might change when learn more about those and try with other complicated recipes like `project.hcl`. For example, project.hcl will interact with a lot of resources, and I don't know if it will work well if we need to encode such heavy templating logic in all three files for every optional field.

Generally, I want to keep `main.tf` and `variables.tf` simpler, reduce the level of templating there, and decouple them more from the schema of tfengine configs since the goal is to make each recipe a more standard and standalone Terraform module and can be used with Private Catalog. Imagine the case where users just use tfengine to generate an initial version of the Terraform configs and want to modify on top afterwards. The optional variables at tfengine level should not be entirely hidden in the Terraform configs, but instead, they should still be made available but just listed as optional. We should try to leverage more existing features of Terraform instead of relying on our templating layer to achieve things like required/optional fields.

I hope eventually the only place that tfengine will write actual values to and do heavy if/else is in the `terraform.tfvars` file which is the standard way of how values are passed to terraform modules.

But again, i'm ok with check in this PR after addressing other comments and using your current strategy.

Reply:
I want to keep `main.tf` and `variables.tf` simple as well. So I agree in general.

I'd like to explore the `optional()` hcl function and other approaches mentioned in another simpler recipe, 'cause things are difficult on this one. Especially because tests on my Pixelbook seem to be significantly slower than in gcb.

I can start testing this config on the private catalog, so I'll tell you what I found

Reply:
sounds good